zilliqa/src/consensus.rs

```diff
@@ -1694,22 +1733,43 @@ impl Consensus {
         Ok(())
     }
 
-    pub fn leader(&self, committee: &Committee, view: u64) -> Validator {
+    pub fn leader_at_block(&mut self, block: &Block, view: u64) -> Validator {
+        let curr_root_hash = self.state.root_hash().unwrap();
+        let block_root_hash = block.state_root_hash();
+
+        self.state.set_to_root(H256(block_root_hash.0));
+        let leader = self.leader(view);
+        self.state.set_to_root(H256(curr_root_hash.0));
+        leader
+    }
+
+    pub fn leader(&mut self, view: u64) -> Validator {
```

Comment:
This `mut` looks unnecessary too - I'm surprised `clippy` doesn't seem to highlight it